### PR TITLE
COMPASS-4451 - Add release scripts to main compass repo

### PR DESCRIPTION
**Pull request by mcasimir**
This PR contains release automation scripts to aid the release tasks from the local machine of developers.

#### `checkout`

``` sh
npm run release checkout <MAJOR.MINOR>
```

Only runnable from master. Checks out (creating it if not existing)  a `MAJOR.MINOR-releases` branch.

``` sh
npm run release checkout 1.22
```

#### `beta`

``` sh
npm run release beta
```

Only runnable from a release branch. It bumps the version to a new beta and pushes it triggering the release.

The new prerelease version is calculated according to the branch name and previous version as follows:

- If the release branch is `1.22` and `package.version < 1.22.0`, it will create `1.22.0-beta.0`.
- When package.version is `1.22.0-beta.0`, do `1.22.0-beta.1`.
- When package.version is `1.22.1` do `1.22.2-beta.0`.

#### `ga`

``` sh
npm run release ga
```

Only runnable from a release branch. It bumps the version to the next ga and pushes it triggering the release.

The new version is calculated according to the branch name and previous version as follows:

- If the release branch is `1.22` and `package.version < 1.22.0`, it will create 1.22.0.
- When package.version is `1.22.0-beta.0`, do `1.22.0`.
- When package.version is `1.22.1` do `1.22.2`.

#### `wait`

``` sh
npm run release wait
```

Only runnable from a release branch. It waits for all the assets of a release to be uploaded.

#### `changelog`

``` sh
npm run release changelog
```

Only runnable from a release branch. Prints the git log between a release and the previous one. Cleans duplicated commits and tag commits.

- If the new release is a GA it will print the log from the previos GA skipping any beta in between.
- If the new release is a beta it will print the log from a previous release of any kind (GA or beta).

#### `publish`

``` sh
npm run release publish
```

Only runnable from a release branch. It completes the release by uploading a new
download center configuration and making sure the associated github release is published.

This command is retryable. Issuing `run release publish` on an older release
will not break newer releases.

**NOTE:** this command requires the following environment variables to be set:

- `MONGODB_DOWNLOADS_AWS_ACCESS_KEY_ID`
- `MONGODB_DOWNLOADS_AWS_SECRET_ACCESS_KEY`

It will perform the following tasks:

1. Make sure that all the assets are reacheable.
2. Downloads and patches the download center configuration with the
   new version. If the old release is >= than the current one skips this step.
3. Waits for a draft github release to be available.
4. Prompts to publish the github release if not published.
5. Waits for the github release to be published.

### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [x] Misc

## Open Questions

Tests in this folder are run independenly:  `npm run test-release-tasks`.

## Types of changes

- [ ] *Backport Needed*
- [x] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Review by rose-m — COMMENTED** [review]
Only two minor comments.

Checked out the code locally and successfully ran all tests ✅

File: release/publish.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    ),
+    'The Github release is still in draft.'
+  );
+  cli.info(`Please review and publish the release on Github:\n\t${release.html_url}`);
+  cli.info('');
+  cli.info('You can run', chalk.bold('npm run release changelog'), 'to get the release notes.');
+  cli.info('');
+
+  cli.action.start('Waiting for Github release to be published');
+  if (!release.draft) {
```
[the remark this answers is not in the document]
Comment (mcasimir) [reply release/publish.js:61]: Oh nice i copy pasted an entire block of code without noticing! Thanks for catching it!

[the remark this answers is not in the document]
Comment (mcasimir) [reply release/publish.js]: Ok i'm more sane that that .. moved it and just forgot to delete it from where it was.

File: package.json [hunk 2]
```diff
@@ -228,14 +228,16 @@
...
+    "release-evergreen": "hadron-build release",
+    "release": "node release/index.js",
     "test-unit": "npm test -- --unit",
     "test-enzyme": "npm test -- --enzyme",
     "test-packages": "npm test -- --packages",
     "test-main": "npm test -- --main",
     "test-renderer": "npm test -- --renderer",
     "test-functional": "npm test -- --functional",
     "test-release": "npm test -- --release",
+    "test-release-tasks": "mocha --timeout 30000 release/*.spec.js",
```
Comment (Anemy) [inline package.json:240]: Do these tests run automatically ever?  Do we want them to?

**Review by Anemy — APPROVED** [review]
Sweeeet - it's really nice having this in this repo and in such a nice understandable way. Nice work.

Do we want these scripts in typescript at some point?

And one note: Once this is in I think we want to update the auto update server docs with the updates to the new release process: https://github.com/10gen/compass-mongodb-com

**Review by rose-m — APPROVED** [review]
Awesome, thanks a lot - having that really makes releasing and working with it a lot easier for newbies like me 😄 👏

File: release/download-center.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  }
+
+  const stringifiedDoc = JSON.stringify(originalVersionDoc);
+  const replaced = stringifiedDoc
+    .replace(
+      new RegExp(_.escapeRegExp(originalVersion), 'g'),
+      releaseVersion
+    )
+    .replace(
+      new RegExp(_.escapeRegExp(originalVersionDeb), 'g'),
```
[the remark this answers is not in the document]
Comment (mcasimir) [reply release/download-center.js:112]: it doesn't, basically what ends up to happening is:

This for `beta`:

``` js
configFileContent
   .replace(/1.2.3-beta/g, '1.2.4-beta')
   .replace(/1.2.3~beta/g, '1.2.4~beta')
```

And for GA is super silly as it replaces the same string twice:

``` js
configFileContent
   .replace(/1.2.3/g, '1.2.4')
   .replace(/1.2.3/g, '1.2.4')
```

File: package.json (hunk 2 above)
Comment (mcasimir) [reply package.json:240]: I have no idea. Hopefully once we switch to a monorepo that will go in its own folder and tests will be run as they are for any other package. I'm not sure i can follow the whole `hadron-build` mechanics yet.

**Pull request merged**